### PR TITLE
test: add unit tests for v0.3.5 modules (triton_kernels, fx_executor, gpu_metrics)

### PR DESCRIPTION
**Pull request by Nickelode00Ns**
## Summary

Adds comprehensive unit tests for the v0.3.5 modules that were previously untested:
- `zenith/runtime/triton_kernels.py`
- `zenith/runtime/fx_executor.py`
- `zenith/observability/gpu_metrics.py`

## Motivation

These modules were added in v0.3.5 but had no corresponding test coverage. This PR addresses that gap with production-grade tests.

### tests/python/test_triton_kernels.py
- **6 test classes, 15+ test methods**
- Tests availability detection and version retrieval
- Tests kernel map generation
- Tests kernel registration with registry
- Tests fused operations fallback (CPU execution without Triton)
- Tests numerical accuracy versus PyTorch reference
- Tests benchmark utility function

### tests/python/test_fx_executor.py
- **8 test classes, 20+ test methods**
- Tests FXExecutionStats dataclass
- Tests FXKernelExecutor initialization and configuration
- Tests dispatcher availability detection
- Tests function wrapping and execution
- Tests statistics tracking and reset
- Tests factory functions
- Tests PyTorch module integration

### tests/python/test_gpu_metrics.py
- **8 test classes, 20+ test methods**
- Tests GPUStats dataclass and serialization
- Tests singleton pattern implementation
- Tests availability detection
- Tests stats collection with mock data
- Tests module-level convenience functions
- Tests thread safety of concurrent access
- Tests memory calculation consistency

## Design Decisions

1. **GPU-Independent Tests**: All tests are designed to run without GPU by testing fallback paths and mocked data
2. **Proper Skip Logic**: Tests requiring GPU use `pytest.skip()` when hardware unavailable
3. **Numerical Accuracy**: Fused operations tested against PyTorch reference with tolerance
4. **Thread Safety**: GPU metrics singleton tested for concurrent access

## Test Count

| File | Classes | Methods |
|------|---------|---------|
| test_triton_kernels.py | 6 | 15+ |
| test_fx_executor.py | 8 | 20+ |
| test_gpu_metrics.py | 8 | 20+ |
| **Total** | **22** | **55+** |

## Running the Tests

```bash
pytest tests/python/test_triton_kernels.py -v
pytest tests/python/test_fx_executor.py -v
pytest tests/python/test_gpu_metrics.py -v
```

Or all at once:
```bash
pytest tests/python/test_triton_kernels.py tests/python/test_fx_executor.py tests/python/test_gpu_metrics.py -v
```

___

### **PR Type**
Tests

___

### **Description**
- Adds comprehensive unit tests for three v0.3.5 modules

- Tests cover triton_kernels, fx_executor, and gpu_metrics

- Includes 22 test classes with 55+ test methods total

- All tests designed to run without GPU via fallback paths

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["triton_kernels module"] --> B["test_triton_kernels.py"]
  C["fx_executor module"] --> D["test_fx_executor.py"]
  E["gpu_metrics module"] --> F["test_gpu_metrics.py"]
  B --> G["6 test classes<br/>15+ methods"]
  D --> H["8 test classes<br/>20+ methods"]
  F --> I["8 test classes<br/>20+ methods"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_triton_kernels.py</strong><dd><code>Triton kernels module unit tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/python/test_triton_kernels.py

<ul><li>Tests Triton availability detection and version retrieval<br> <li> Tests kernel map generation and registration with mock registry<br> <li> Tests fused operations fallback to PyTorch when Triton unavailable<br> <li> Tests numerical accuracy of fused_linear_gelu and fused_linear_relu <br>against PyTorch reference<br> <li> Tests benchmark utility function with proper error handling</ul>

</details>

  </td>
  <td><a href="https://github.com/vibeswithkk/ZENITH/pull/9/files#diff-5bc10501252a6abafba1635b00e532ffac991367e98e0e757a492cd611a4214d">+248/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_fx_executor.py</strong><dd><code>FX executor module unit tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/python/test_fx_executor.py

<ul><li>Tests FXExecutionStats dataclass initialization and field mutability<br> <li> Tests FXKernelExecutor initialization with custom precision and device <br>parameters<br> <li> Tests dispatcher availability detection and function wrapping <br>functionality<br> <li> Tests statistics tracking, reset, and factory functions<br> <li> Tests integration with PyTorch tensors and nn.Module forward methods</ul>

</details>

  </td>
  <td><a href="https://github.com/vibeswithkk/ZENITH/pull/9/files#diff-0225ffb50489df798ddb1b49b47b517673cabd324f8c6c006c0507a77c08026f">+309/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_gpu_metrics.py</strong><dd><code>GPU metrics collector module unit tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/python/test_gpu_metrics.py

<ul><li>Tests GPUStats dataclass creation with required and optional fields<br> <li> Tests singleton pattern implementation and reset functionality<br> <li> Tests availability detection and device count retrieval<br> <li> Tests stats collection with invalid device indices and mocked pynvml<br> <li> Tests module-level convenience functions and thread safety of <br>concurrent access<br> <li> Tests memory calculation consistency and utilization range validation</ul>

</details>

  </td>
  <td><a href="https://github.com/vibeswithkk/ZENITH/pull/9/files#diff-c43e506dc3a867ee68c37135aa84be2b6f68ec51980f0bbe1c1c598338a34ff5">+343/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by Nickelode00Ns** [comment]
Thank you for the detailed review, Wahyu.

**1. GPU-Independent Test Design Approach**

The design follows the principle of testing the contract, not the implementation. Here is the methodology:

a) **Layered Testing Strategy**
- Public API layer tests (is_available, get_version, etc.) verify return types and interface contracts
- These tests pass regardless of hardware availability since they test the API behavior, not GPU execution

b) **Conditional Skip Pattern**
- Tests requiring actual GPU execution use pytest fixtures that detect hardware availability
- When GPU is unavailable, tests skip gracefully rather than fail
- Example: torch_available fixture checks for PyTorch import before running tensor operations

c) **Fallback Path Coverage**
- The modules themselves have fallback paths (e.g., fused_linear_gelu falls back to PyTorch native ops)
- Tests verify these fallback paths work correctly on CPU
- This ensures CI passes on standard GitHub runners without GPU

d) **Mock-Based Validation**
- For GPU metrics, we create mock GPUStats objects to verify serialization and data flow
- This tests the data handling logic without requiring actual NVML calls

**2. Challenges Encountered**

a) **Import Chain Complexity**: The modules have conditional imports (Triton, pynvml, torch). Ensuring tests handle all import failure scenarios required careful fixture design.

b) **Singleton Reset**: GPUMetricsCollector uses singleton pattern. Tests needed explicit reset() calls between test cases to ensure isolation.

c) **Numerical Tolerance**: Fused kernel outputs may differ slightly from PyTorch reference due to floating-point operation ordering. Used torch.testing.assert_close with appropriate rtol/atol values.

**Suggested Documentation Improvements**

Based on these challenges, the following documentation additions would help future contributors:
1. A TESTING.md guide explaining the conditional skip patterns
2. Docstrings in modules explaining fallback behavior
3. Example test fixtures in conftest.py for hardware detection

I am happy to create these documentation additions in a follow-up PR if useful.

Best regards,
Nickelode00Ns

**Comment by vibeswithkk** [comment]
Excellent breakdown, @Nickelode00Ns. Your testing methodology is exactly what we need - testing contracts rather than implementations is the right approach for a library that needs to work across different hardware configurations.

I particularly appreciate:
- The layered testing strategy explanation
- The singleton reset pattern for test isolation
- The numerical tolerance considerations

**Regarding your suggested documentation improvements:**

Yes, please proceed with a follow-up PR for TESTING.md. This will be valuable for future contributors. Include:
1. Conditional skip patterns
2. Hardware detection fixture examples
3. Guidelines for fallback path testing

Merging this PR now. Welcome to the team.

Wahyu
